### ISSUE TITLE
Serverpool does not seem to use the second server

### ISSUE THREAD

**Issue by cornelinux** [cannatag/ldap3#649]
I am observing an unexpected behaviour with a server pool.
There are two server in the server pool

The server pool is created like this:

~~~~python
server_pool = ldap3.ServerPool(None, ldap3.ROUND_ROBIN, active=2, exhaust=30)
# then we add two servers like this:
server = ldap3.Server(host, port=port, use_ssl=ssl, connect_timeout=5, get_info=get_info, tls=tls_context)
server_pool.add(server)
~~~~

To my understanding requests to to the two servers and if a server does not response within 5 seconds (``connect_timeout``) the server will be removed from the pool for 30 seconds.

We start a connection like this:

~~~~python
ldap3.Connection(server_pool, user=user, password=password,
                             auto_bind=False,
                                 client_strategy=ldap3.SYNC,
                                 authentication=ldap3.simple,
                                 check_names=True,
                                 auto_referrals=False)
~~~~
Note: We do currently not use the ``receive_timeout``. 

"Every now and then"(TM) we see an "LDAPSocketOpenError". One of the servers seems to be removed from the pool but it looks like the 2nd server is not queried. 

I would expect that if one server is removed from the pool the next server is used?
What can possibly go wrong with timeouts?

(I am sorry this is part of a more complicated code, which is not simply postable. So I first wanted to make sure, if my understanding and epxectation is right)

**Comment by fredreichbier** [comment]
(I work with @cornelinux)

I did a dive into the ``ServerPool`` code and have some questions. :-) If I understand correctly, a ``ServerPoolState`` object stores what servers are currently available and what servers are down. However, the ``ServerPoolState`` is stored *for each connection separately*, which means that if we create a second SYNC connection with the same ``ServerPool``, the second connection does not have the information about unavailable servers from the ``ServerPoolState`` of the first connection, and potentially tries them all again. Did I understand this correctly? :-)

I read the related issue #465 in which you recommend using the REUSABLE strategy. I'm not sure if we want to use that strategy -- in case we don't want to, I'm thinking about implementing our own ``ServerPool`` which holds the information about available and unavailable servers *centrally*. Do you think this is, in principle, possible, or am I overlooking something? (Of course, we would have to take special care about threading and synchronization).

Thank you for all your help!

**Comment by cannatag** [comment]
Hi, sorry for the late answer. If you assign the same ServerPool to different connections it should be shared by all connections. Is this working as you expect?

The Reusable strategy is more complex. It keeps a pool of connections open an use an asynchronous strategy for communicating with many server at the same time. I don't think this is what you need.

Let me know of any problem you still have.

Bye,
Giovanni

**Comment by fredreichbier** [comment]
Hi Giovanni, thank you for your reply and the pointers!

> The Reusable strategy is more complex. It keeps a pool of connections open an use an asynchronous strategy for communicating with many server at the same time. I don't think this is what you need.

Thanks for the hint! In that case, I guess the reusable strategy really isn't what we need.

> If you assign the same ServerPool to different connections it should be shared by all connections. Is this working as you expect?

Hmm, the ServerPool isn't really working how I would expect it to work. I put together a test script. It creates a ``ServerPool`` with 5 servers, of which only one serves LDAP (the FreeIPA demo server). The other four point to unoccupied IP addresses. It then creates two connections from this server pool:
```python
import sys

import logging
logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)

import ldap3
from ldap3.utils.log import set_library_log_activation_level, set_library_log_detail_level, NETWORK

set_library_log_activation_level(logging.DEBUG)
set_library_log_detail_level(NETWORK)

pool = ldap3.ServerPool(None, ldap3.ROUND_ROBIN, active=True, exhaust=30)

for host in ['ipa.demo1.freeipa.org', '192.0.2.100', '192.0.2.101', '192.0.2.102', '192.0.2.103']:
    server = ldap3.Server(host, connect_timeout=5)
    pool.add(server)

logging.info("Connection #1")
ldap3.Connection(pool, auto_bind=True)

logging.info("Connection #2")
ldap3.Connection(pool, auto_bind=True)
```

I use the ldap3 logging facilities to find out what servers are checked for availability. Here is an example run (individual runs differ because the round robin strategy chooses a random initial server):
```console
$ python2 test-serverpool.py| grep -e ':root:' -e 'server <'       
INFO:root:Connection #1
DEBUG:ldap3:NETWORK:checking server <ldap://192.0.2.102:389 - cleartext> for availability
DEBUG:ldap3:ERROR:server <ldap://192.0.2.102:389 - cleartext> not available at <[2, 1, 6, '', ('192.0.2.102', 389), False, datetime.datetime(2019, 2, 11, 14, 47, 16, 606889)]>
DEBUG:ldap3:NETWORK:checking server <ldap://192.0.2.103:389 - cleartext> for availability
DEBUG:ldap3:ERROR:server <ldap://192.0.2.103:389 - cleartext> not available at <[2, 1, 6, '', ('192.0.2.103', 389), False, datetime.datetime(2019, 2, 11, 14, 47, 21, 610048)]>
DEBUG:ldap3:NETWORK:checking server <ldap://ipa.demo1.freeipa.org:389 - cleartext> for availability
DEBUG:ldap3:BASIC:server <ldap://ipa.demo1.freeipa.org:389 - cleartext> available at <[2, 1, 6, '', ('52.57.162.88', 389), None, None]>
INFO:root:Connection #2
DEBUG:ldap3:NETWORK:checking server <ldap://192.0.2.101:389 - cleartext> for availability
DEBUG:ldap3:ERROR:server <ldap://192.0.2.101:389 - cleartext> not available at <[2, 1, 6, '', ('192.0.2.101', 389), False, datetime.datetime(2019, 2, 11, 14, 47, 27, 46967)]>
DEBUG:ldap3:NETWORK:checking server <ldap://192.0.2.102:389 - cleartext> for availability
DEBUG:ldap3:ERROR:server <ldap://192.0.2.102:389 - cleartext> not available at <[2, 1, 6, '', ('192.0.2.102', 389), False, datetime.datetime(2019, 2, 11, 14, 47, 32, 53612)]>
DEBUG:ldap3:NETWORK:checking server <ldap://192.0.2.103:389 - cleartext> for availability
DEBUG:ldap3:ERROR:server <ldap://192.0.2.103:389 - cleartext> not available at <[2, 1, 6, '', ('192.0.2.103', 389), False, datetime.datetime(2019, 2, 11, 14, 47, 37, 60157)]>
DEBUG:ldap3:NETWORK:checking server <ldap://ipa.demo1.freeipa.org:389 - cleartext> for availability
DEBUG:ldap3:BASIC:server <ldap://ipa.demo1.freeipa.org:389 - cleartext> available at <[2, 1, 6, '', ('52.57.162.88', 389), None, None]>
```
We see that the first connection checks 192.0.2.102 and 192.0.2.103 and notes that they're unavailable. So far, everything is working as expected. But we then see that the second connection checks 192.0.2.102 and 192.0.2.103 again, even though the prior connection already noted that they're unavailable! I would expect that they are both skipped, because they have been temporarily removed from the server pool.

I think this situation happens because the set of active servers is stored in a ``ServerPoolState`` object, and every connection instantiates a *new* ``ServerPoolState`` object:
<pre>
$ python2 test-serverpool.py| grep -e ':root:' -e 'ServerPoolState'
INFO:root:Connection #1
DEBUG:ldap3:BASIC:instantiated ServerPoolState: &lt;&lt;ldap3.core.pooling.ServerPoolState object at 0x7f13128b7<b>3</b>50&gt;&gt;
INFO:root:Connection #2
DEBUG:ldap3:BASIC:instantiated ServerPoolState: &lt;&lt;ldap3.core.pooling.ServerPoolState object at 0x7f13128b7<b>9</b>50&gt;&gt;
</pre>
As the two connections hold different ``ServerPoolState`` objects, the information about active and inactive servers is not shared between the connections.

I would expect that the two connections share a ``ServerPoolState`` object instead. But maybe my idea of the server pool is a misconception :-) What do you think?

**Comment by cannatag** [comment]
Hi, sorry for the long delay. Can you check the code in dev? I've added the new parameter ``single`` in the ServerPool definition with default to ``True``. When it's True it use a single state for all connections using the serverpool, when it's False each connection has its own pool state (this is the previous behaviour).

Let me know if this is good for you.

Bye,
Giovanni

**Comment by cannatag** [comment]
Fixed in version 2.6
